ftdc/custom_format.go

```diff
@@ -0,0 +1,549 @@
+// represent 8 metrics.
+func readDiffBits(reader *bufio.Reader, schema *Schema) []int {
+	// 1 diff bit per metric + 1 bit for the packed "schema bit".
+	numBits := len(schema.fieldOrder) + 1
+
+	// If numBits < 8 then numBytes = 1,
+	// ElseIf numBits < 16 then numBytes = 2,
+	// ElseIf numBits < 24 then numBytes = 3, etc...
+	numBytes := 1 + ((numBits - 1) / 8)
+
+	diffBytes := make([]byte, numBytes)
+	_, err := io.ReadFull(reader, diffBytes)
+	if err != nil {
+		panic(err)
+	}
+
+	var ret []int
+	for fieldIdx := 0; fieldIdx < len(schema.fieldOrder); fieldIdx++ {
+		// The 0th metric is addressed via the 1st bit. This is due to the shifting caused by the
+		// packed schema bit.
+		bitIdx := fieldIdx + 1
+
+		// Divide by eight bits per byte to find the offset into the `diffBytes` array.
+		diffByteOffset := bitIdx / 8
+
+		// And take the remainder to address the indidivual bit.
+		bitOffset := bitIdx % 8
+
+		if bitValue := diffBytes[diffByteOffset] & (1 << bitOffset); bitValue > 0 {
+			// The bit is `1`, add the correlated `fieldIdx` to the return value of metrics that
+			// changed.
+			ret = append(ret, fieldIdx)
+		}
+	}
+
+	return ret
+}
+
+// readData returns the "hydrated" metrics for a data reading. For example, if there are ten metrics
+// and none of them changed, the returned []float32 will be identical to `prevValues`. `prevValues`
+// is the post-hydration list and consequently matches the `schema.fieldOrder` size.
+func readData(reader *bufio.Reader, schema *Schema, diffedFields []int, prevValues []float32) ([]float32, error) {
+	if prevValues != nil && len(prevValues) != len(schema.fieldOrder) {
+		return nil, fmt.Errorf("Parser error. Mismatched `prevValues` and schema size. PrevValues: %d Schema: %d",
+			len(prevValues), len(schema.fieldOrder))
+	}
+
+	var ret []float32
+
+	// For each metric in the schema:
+	for dataIdx := 0; dataIdx < len(schema.fieldOrder); dataIdx++ {
+
+		// See if the metric index exists in the `diffedFields` array.
+		diffFromPrev := false
+		for _, fieldIdx := range diffedFields {
+			if dataIdx == fieldIdx {
+				diffFromPrev = true
+				break
+			}
+		}
+
+		if diffFromPrev {
+			// If the metric existed, it's because there was a fresh reading in the input
+			// `reader`. Parse the value from the `reader`.
+			ret = append(ret, 0.0)
+			binary.Read(reader, binary.BigEndian, &ret[dataIdx])
+		} else {
+			// Otherwise, the metric did not change. Use the previous value.
+			if prevValues == nil {
+				// The parser and writer agree that the `prevValues` is `0.0` for all metrics
+				// following a schema change.
+				ret = append(ret, 0.0)
+			} else {
+				ret = append(ret, prevValues[dataIdx])
+			}
+		}
+	}
+
+	return ret, nil
+}
+
+// Hydrate takes the input []float slice of `data` and matches those to their corresponding metric
+// names. Returning a
+func (schema *Schema) Hydrate(data []float32) map[string]any {
```

Comment:
Given that the code right now only returns floats & requires that all metrics have field, why is the return type not:
```
`map[string]map[string]float32`
```
?

Reply:
I waffled on that. I expect (based on prior experience) we might have a "stats" of:
```
type Networking struct {
  WebRTC struct {
    NumPeerConnections int
...
  }
  Http2 struct {
    RequestsServices int
  }
}
```

The reflection right now to walk embedded structures won't work with that, but the types chosen in this PR are reflective of this (admittedly unstated) longer-term vision.